instance/models/mixins/secret_keys.py

```diff
@@ -88,17 +88,19 @@ def get_secret_key_settings(self):
         Render the secret keys template as YAML and return it for use
         in the appserver.
         """
+        key_provider = self.get_key_provider()
         if not self.secret_key_b64encoded:
+            logger.warning('Instance %s does not have a secret key; not prefilling variables.', self)
             return ''
         template = loader.get_template('instance/ansible/secret-keys.yml')
-        return template.render({"secret_generator": self.get_key_provider()})
+        return template.render({"secret_generator": key_provider})
 
     def get_key_provider(self):
         """
         Get a proxy we can pass to a template that'll generate secret keys
         for any attribute requested from it.
         """
-        if not self.appserver_set.count() and not self.secret_key_b64encoded:
+        if not self.appserver_set.exists() and not self.secret_key_b64encoded:
             raise ValueError(
```

Comment:
There should also be a new unit test like `test_secret_key_creation` that creates an instance, updates it to have `secret_key_b64encoded = ""`, then asserts that provisioning an AppServer using `instance.spawn_appserver()` (or just `_create_owned_appserver` if that's easier to reduce the amount of mocking required) fails with this specific `ValueError`.